File: src/world/Grid.js
import { GRID_SIZE } from '../core/Constants.js';
import { Tile } from './Tile.js';

export class Grid {
  constructor() {
    this.size = GRID_SIZE;
    this.tiles = [];
    for (let row = 0; row < this.size; row++) {
      this.tiles[row] = [];
      for (let col = 0; col < this.size; col++) {
        this.tiles[row][col] = new Tile(col, row);
      }
    }
  }

  inBounds(col, row) {
    return col >= 0 && col < this.size && row >= 0 && row < this.size;
  }

  getTile(col, row) {
    if (!this.inBounds(col, row)) return null;
    return this.tiles[row][col];
  }

  isAreaFree(col, row, width, height) {
    for (let r = row; r < row + height; r++) {
      for (let c = col; c < col + width; c++) {
        if (!this.inBounds(c, r)) return false;
        if (this.tiles[r][c].occupiedBy !== null) return false;
      }
    }
    return true;
  }

  occupyArea(col, row, width, height, buildingId) {
    for (let r = row; r < row + height; r++) {
      for (let c = col; c < col + width; c++) {
        const tile = this.tiles[r][c];
        tile.occupiedBy = buildingId;
        tile.walkable = false;
      }
    }
  }

  freeArea(col, row, width, height) {
    for (let r = row; r < row + height; r++) {
      for (let c = col; c < col + width; c++) {
        const tile = this.tiles[r][c];
        tile.occupiedBy = null;
        tile.walkable = true;
      }
    }
  }

  getNeighbors(col, row) {
    const dirs = [
      [-1, 0], [1, 0], [0, -1], [0, 1],
      [-1, -1], [-1, 1], [1, -1], [1, 1],
    ];
    const result = [];
    for (const [dc, dr] of dirs) {
      const nc = col + dc;
      const nr = row + dr;
      if (this.inBounds(nc, nr)) {
        result.push(this.tiles[nr][nc]);
      }
    }
    return result;
  }

  getWalkableNeighbors(col, row) {
    return this.getNeighbors(col, row).filter(t => t.walkable);
  }
}

File: ios/BiteDefense/Core/Grid.swift
import Foundation

/// Tile-occupancy grid. Mirrors the occupancy half of `Grid.js` — separate
/// from the visual tile map (`IsoTileMap`) so systems can reason about
/// placement without needing the rendering layer.
final class Grid {
    let cols: Int
    let rows: Int
    /// `nil` = empty; otherwise the building ID occupying that tile.
    private var occupancy: [[Int?]]

    init(cols: Int = Constants.gridCols, rows: Int = Constants.gridRows) {
        self.cols = cols
        self.rows = rows
        self.occupancy = Array(repeating: Array(repeating: nil, count: cols),
                               count: rows)
    }

    func inBounds(col: Int, row: Int) -> Bool {
        (0..<cols).contains(col) && (0..<rows).contains(row)
    }

    func isAreaFree(col: Int, row: Int, width: Int, height: Int,
                    ignoring buildingId: Int? = nil) -> Bool {
        for r in row..<(row + height) {
            for c in col..<(col + width) {
                guard inBounds(col: c, row: r) else { return false }
                if let id = occupancy[r][c], id != buildingId { return false }
            }
        }
        return true
    }

    func occupy(col: Int, row: Int, width: Int, height: Int, buildingId: Int) {
        for r in row..<(row + height) {
            for c in col..<(col + width) where inBounds(col: c, row: r) {
                occupancy[r][c] = buildingId
            }
        }
    }

    func free(col: Int, row: Int, width: Int, height: Int) {
        for r in row..<(row + height) {
            for c in col..<(col + width) where inBounds(col: c, row: r) {
                occupancy[r][c] = nil
            }
        }
    }

    func buildingId(at col: Int, row: Int) -> Int? {
        guard inBounds(col: col, row: row) else { return nil }
        return occupancy[row][col]
    }

    /// Wipe all occupancy. Called during save-load to rebuild from models.
    func clear() {
        occupancy = Array(repeating: Array(repeating: nil, count: cols), count: rows)
    }
}
